thread_pool.c

```diff
@@ -874,13 +905,35 @@ int hts_tpool_process_flush(hts_tpool_process *q) {
     if (q->qsize < q->n_output + q->n_input + q->n_processing)
         q->qsize = q->n_output + q->n_input + q->n_processing;
 
-    // Wait for n_input and n_processing to hit zero.
-    while (q->n_input || q->n_processing) {
-        while (q->n_input)
-            pthread_cond_wait(&q->input_empty_c, &p->pool_m);
-        if (q->shutdown) break;
+    // When shutdown, we won't be launching more, but we can still
+    // wait for any processing jobs complete.
+    if (q->shutdown) {
         while (q->n_processing)
             pthread_cond_wait(&q->none_processing_c, &p->pool_m);
+    }
+
+    // Wait for n_input and n_processing to hit zero.
+    while (!q->shutdown && (q->n_input || q->n_processing)) {
+        struct timeval now;
+        struct timespec timeout;
+
+        while (q->n_input && !q->shutdown) {
+            gettimeofday(&now, NULL);
+            timeout.tv_sec = now.tv_sec + 1;
+            timeout.tv_nsec = now.tv_usec * 1000;
+            pthread_cond_timedwait(&q->input_empty_c, &p->pool_m, &timeout);
```

Comment:
Is this using timedwait to avoid a lock-up found in testing, or is this just a belt-and-braces approach?

Reply:
If I recall it was something I found while testing.  However given multiple rounds of testing and fixing, that doesn't necessarily mean it is still needed.  Probably, but with several subsequent changes it's hard to be sure without removing and testing again.